### PR TITLE
Implement JUnit5 support and parameterize tests

### PR DESCRIPTION
**Pull request by blorente**
# Checklist

- [X] I have filed an issue about this change and discussed potential changes with the maintainers.
- [X] I have received the approval from the maintainers to make this change.
- [ ] This is not a stylistic, refactoring, or cleanup change.
  - Well, this has some refactoring, mostly of the test code to make it easier to parameterize.

Please note that the maintainers will not be reviewing this change until all checkboxes are ticked. See 
the [Contributions](https://github.com/bazelbuild/intellij#contributions) section in the README for more 
details.

# Discussion thread for this change

Issue number: #4433

# Description of this change

Luckily, JUnit 4 test filters are compatible with JUnit 5.

Therefore, the only things that this PR needs to do are:
- Change `JUnitTestHeuristic` to account for JUnit 5.
- Change the test filter generators to account for JUnit 5 as well.
- Change the testbed so that it also tests against JUnit 5 code:
  - Introduce `JUnitTestUtils`, a utility class to set up test repositories with JUnit 5 or 4 code.
  - Introduce `BlazeJUnitRunConfigurationProducerTestCase`, to remove parameterization boilerplate for the Run Configuration Producer tests.
  - Parameterize relevant integration tests (either with the base class mentioned above, or by hand), to test a set of JUnit versions (currently, 4 and 5). This mostly affects the set up of the code under tests, so we can reuse a lot of the test logic.
  - Introduce JUnit 5 to the java example project.

- Implement fastbuild support for `java_junit5_test`:
  - Add the `java_junit5_test` as an allowed target kind for FastBuild.
  - Capture the `main_class` attribute in the FastBuild aspect proto.
  - Prefer `main_class` when building run configurations over the default bazel runner (`java_junit5_test` uses a custom test runner).

### CODE REVIEW

**Comment by raoarun** [comment]
@mai93 please review.

**Comment by blorente** [comment]
Rebased on top of `master` and implemented https://github.com/bazelbuild/intellij/issues/5550 in the last two commits.

**Comment by kermitnirmit** [comment]
Hi - I'm super interesting in this because I use junit5 and my parameterized tests don't look like this: 
image
They are all on the top level instead. I would like this hierarchy.

File: java/tests/unittests/com/google/idea/blaze/java/run/producers/BlazeJUnitTestFilterFlagsTest.java [hunk 1]
```diff
@@ -63,6 +71,25 @@ public void testParameterizedIgnoredForSingleClass() {
         .isEqualTo("com.google.idea.ClassName#");
   }
 
+  @Test
+  public void testParameterizedIgnoredForSingleClassJUnit5() {
+    assertThat(
```
Comment (mai93) [inline java/tests/unittests/com/google/idea/blaze/java/run/producers/BlazeJUnitTestFilterFlagsTest.java:76]: what is the difference between this test and `testSingleJUnit5ClassFilter`?

File: java/tests/integrationtests/com/google/idea/blaze/java/run/producers/MultipleJavaClassesTestContextProviderTest.java [hunk 2]
```diff
@@ -71,63 +77,32 @@ public final void addSourceFolder() {
...
   @After
   public final void removeSourceFolder() {
+    VirtualFile pkgRoot = workspace.createDirectory(pkgRootPath);
     ApplicationManager.getApplication()
         .runWriteAction(
             () -> {
               final ModifiableRootModel model =
                   ModuleRootManager.getInstance(testFixture.getModule()).getModifiableModel();
               ContentEntry contentEntry = model.getContentEntries()[0];
-              contentEntry.clearSourceFolders();
```
[the remark this answers is not in the document]
Comment (blorente) [reply java/tests/integrationtests/com/google/idea/blaze/java/run/producers/MultipleJavaClassesTestContextProviderTest.java:101]: Good question, it took me a while to find the answer 😓 .
I've left it in a comment, but the reason is:

```
// There is a global content entry, `//src`, which is necessary to infer targets from test classes.
// This is set outside of these test (somewhere in the superclass) once per run, not once per test.
// If we just use contentEntry.clearSourceDirs() here, it won't be re-created, and tests that assume it exists will fail.
// Therefore, we only delete the content entries that this test has created.
```

File: java/tests/unittests/com/google/idea/blaze/java/run/producers/BlazeJUnitTestFilterFlagsTest.java (hunk 1 above)
Comment (blorente) [reply java/tests/unittests/com/google/idea/blaze/java/run/producers/BlazeJUnitTestFilterFlagsTest.java:76]: You're absolutely right, there's no difference. That means the test above it (the same, but for junit4) is broken as well. Will fix them both and make them meaningful.

**Comment by blorente** [comment]
Addressed the comments, sorry for the delay but it was hard to find time, as per usual.

**Review by agluszak — APPROVED** [review]

**Pull request merged**